packages/flutter_tools/lib/src/build_system/targets/web.dart

```diff
@@ -732,3 +735,281 @@ function onlineFirst(event) {
+    final String jsNameWithMd5 = _getFileNameFromPath(md5JsPath);
+    final String sourceMapName = '$jsName.map';
+    final String sourceMapNameWithHash = '$jsNameWithMd5.map';
+
+    final File? jsFile = _convertFullPathToFile(environment, noMd5JsPath);
+    if (jsFile != null) {
+      final String jsContent = jsFile.readAsStringSync();
+      final String newJsContent = jsContent.replaceAll(sourceMapName, sourceMapNameWithHash);
+      jsFile.writeAsStringSync(newJsContent);
+    }
+    // sourceMapFile.rename('$md5JsPath.map');
+    if (sourceMapFile != null && sourceMapFile.existsSync()) {
+      sourceMapFile.copy('$md5JsPath.map');
+    }
+  }
+
+  /// xxx.part.js add hash, for example main.dart.js_1.part.js --> main.dart.js_1.part.[hash].js
+  Future<String> jsAddHash(Environment environment) async{
+    final List<String> list = resourcesMap['js'] ?? <String>[];
+    final List<String> partJsPathArr = <String>[];
+    String mainJsPath = '';
+    String mainJsContent = '';
+    final RegExp partReg = RegExp(r'main\.(.+)\.part\.js$');
+    final RegExp mainReg = RegExp(r'main.dart.js$');
+    String newMainJsPath = '';
+    String mainHash = '';
+    /// filter main.dart.js and xxx.part.js
+    for (final String path in list) {
+      if (partReg.hasMatch(path)) {
+        partJsPathArr.add(path);
+      } else if (mainReg.hasMatch(path)) {
+        mainJsPath = path;
+      }
+    }
+    // final File mainJsFile = globals.fs.file(mainJsPath);
+    final File? mainJsFile = _convertFullPathToFile(environment, mainJsPath);
+    if (mainJsFile != null) {
+      mainHash = await _getFileMd5(mainJsFile);
+      newMainJsPath = renameFileName(mainJsPath, mainHash);
+      // main.dart.js add hash
+      _sourceMapAddHash(environment, mainJsPath, newMainJsPath, mainHash);
+      mainJsContent = mainJsFile.readAsStringSync();
+    }
+
+
+    /// craete xxx.part.[hash].js and replace new name in main.dart.js
+    /// 1. rename part.js with hash
+    /// 2. replace xxx.part.[hash].js in main.dart.js
+    /// 3. modify sourcemap in part.js
+    Future<void> _dealPartJs(String partJsPath) async {
+      // final File partJsFile = globals.fs.file(partJsPath);
+      final File? partJsFile = _convertFullPathToFile(environment, partJsPath);
+      if (partJsFile != null) {
+        final String partHash = await _getFileMd5(partJsFile);
+        final String newPartJsPath = renameFileName(partJsPath, partHash);
+        final String oldPartJsName = _getFileNameFromPath(partJsPath);
+        final String newPartJsName = _getFileNameFromPath(newPartJsPath);
+        // modify sourcemap in xxx.part.js
+        _sourceMapAddHash(environment, partJsPath, newPartJsPath, partHash);
+        // rename xxx.part.js with hash
+        await partJsFile.rename(newPartJsPath);
+        // replace xxx.part.[hash].js in main.dart.js
+        mainJsContent = mainJsContent.replaceAll(oldPartJsName, newPartJsName);
```

Comment:
This is not robust. It can accidentally replace unintended strings. We should work with the compiler team to allow customizing part file URLs.

@sigmundch Do you think it would be possible to have the main bundle provide API for customizing loading of part files?

Reply:
To your direct question - there is the [dartDeferredLibraryLoader](https://github.com/dart-lang/sdk/blob/8614f86c9b1aa947766ca7f6ed34bf37a8670d02/pkg/compiler/lib/src/js_emitter/headers.dart#L30) JavaScript hook. This could be leveraged to adjust the uri and override how to download the resource. That said, I also feel that this would be handled more robustly externally without renaming file names, but using the hash in URL parameters or paths.

One suggestion, if I may, would be to consider all .js files together in one cache group. Given the whole-program nature of dart2js, I'd expect that small modifications in the Dart sources could reflect in invalidations of multiple .js files. Also, the main.dart.js file already has in it's contents a md5 hash for each of the part files, so you'll see it change on most modifications. Technically you could say that a .part file could be cached, even when the main.dart.js file is not, but I doubt there will be much benefit in doing so given that the main.dart.js file is the most relevant in the initial page load of an app.

One benefit of treating all .js in one group is that you could version all .js sources using a single hash and make it a path prefix:
```html
  <script src="a414f03/main.dart.js"></script>
```